src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/DistributedTM/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,296 @@
+        {
+            if (this.key.ETag != expectedETag)
+                throw new ArgumentException(nameof(expectedETag), "Etag does not match");
+
+            // assemble all storage operations into a single batch
+            // these operations must commit in sequence, but not necessarily atomically
+            // so we can split this up if needed
+            var batchOperation = new TableBatchOperation();
+
+            // first, clean up aborted records
+            if (abortAfter.HasValue && states.Count != 0)
+            {
+                while (states.Count > 0 && states[states.Count - 1].Key > abortAfter)
+                {
+                    var entity = states[states.Count - 1].Value;
+                    batchOperation.Delete(entity);
+                    states.RemoveAt(states.Count - 1);
+
+                    if (logger.IsEnabled(LogLevel.Trace))
+                        logger.LogTrace($"{partition}.{states[states.Count - 1].Key:x16} Delete {entity.TransactionId}");
+                }
+            }
+
+            // second, persist non-obsolete prepare records
+            var obsoleteBefore = commitUpTo.HasValue ? commitUpTo.Value : key.CommittedSequenceId;
+            if (statesToPrepare != null)
+                foreach (var s in statesToPrepare)
+                    if (s.SequenceId >= obsoleteBefore)
+                    {
+                        if (FindState(s.SequenceId, out var pos))
+                        {
+                            // overwrite with new pending state
+                            var existing = states[pos].Value;
+                            existing.TransactionId = s.TransactionId;
+                            existing.TransactionTimestamp = s.TimeStamp;
+                            existing.TransactionManager = s.TransactionManager;
+                            existing.SetState(s.State, this.jsonSettings);
+                            batchOperation.Replace(existing);
+                            states.RemoveAt(pos);
+
+                            if (logger.IsEnabled(LogLevel.Trace))
+                                logger.LogTrace($"{partition}.{existing.SequenceId:x16} Update {existing.TransactionId}");
+                        }
+                        else
+                        {
+                            var entity = StateEntity.Create(this.jsonSettings, this.partition, s);
+                            batchOperation.Insert(entity);
+                            states.Insert(pos, new KeyValuePair<long, StateEntity>(s.SequenceId, entity));
+
+                            if (logger.IsEnabled(LogLevel.Trace))
+                                logger.LogTrace($"{partition}.{s.SequenceId:x16} Insert {entity.TransactionId}");
+                        }
+                    }
+
+            // third, persist metadata and commit position
+            key.Metadata = metadata;
```

Comment:
the key serves as a synchronizer, to prevent modification by multiple grains under edge conditions, like duplicate activations or deployments.  Every batch write needs to include the key, even if the key values don't change.